File: java-service/src/main/java/com/edu/platform/model/individual/IndividualLessonTopic.java
package com.edu.platform.model.individual;

import com.edu.platform.model.StudentProfile;
import com.edu.platform.model.Subject;
import com.edu.platform.model.Term;
import jakarta.persistence.*;
import lombok.*;

import java.math.BigDecimal;
import java.time.Instant;

/**
 * Lesson topics extracted from INDIVIDUAL student scheme of work documents.
 * These are unique to each INDIVIDUAL student, unlike shared lesson_topics table.
 */
@Entity
@Table(name = "individual_lesson_topics", schema = "academic")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class IndividualLessonTopic {
    
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    
    /**
     * The student this topic belongs to
     */
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "student_profile_id", nullable = false)
    private StudentProfile studentProfile;
    
    /**
     * The scheme this topic was extracted from
     */
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "scheme_id", nullable = false)
    private IndividualStudentScheme scheme;
    
    /**
     * Subject as identified in the uploaded document
     */
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "subject_id", nullable = false)
    private Subject subject;
    
    // ============================================================
    // TOPIC DETAILS
    // ============================================================
    
    @Column(name = "topic_title", nullable = false, length = 255)
    private String topicTitle;
    
    @Column(name = "description", columnDefinition = "TEXT")
    private String description;
    
    @Column(name = "week_number")
    private Integer weekNumber;
    
    // ============================================================
    // SUBJECT MAPPING (AI-generated)
    // ============================================================
    
    /**
     * Platform subject this topic was mapped to (can be different from extracted subject)
     */
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "mapped_subject_id")
    private Subject mappedSubject;
    
    /**
     * Confidence in the subject mapping (0.00 to 100.00)
     */
    @Column(name = "mapping_confidence", precision = 5, scale = 2)
    private BigDecimal mappingConfidence;
    
    // ============================================================
    // LEARNING RESOURCES
    // ============================================================
    
    @Column(name = "file_name", length = 255)
    private String fileName;
    
    @Column(name = "file_url", length = 500)
    private String fileUrl;
    
    // ============================================================
    // ACADEMIC CONTEXT
    // ============================================================
    
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "term_id")
    private Term term;
    
    // ============================================================
    // TIMESTAMPS
    // ============================================================
    
    @Column(name = "created_at", nullable = false, updatable = false)
    private Instant createdAt;
    
    @Column(name = "updated_at", nullable = false)
    private Instant updatedAt;
    
    // ============================================================
    // LIFECYCLE CALLBACKS
    // ============================================================
    
    @PrePersist
    protected void onCreate() {
        Instant now = Instant.now();
        if (this.createdAt == null) {
            this.createdAt = now;
        }
        if (this.updatedAt == null) {
            this.updatedAt = now;
        }
    }
    
    @PreUpdate
    protected void onUpdate() {
        this.updatedAt = Instant.now();
    }
    
    // ============================================================
    // HELPER METHODS
    // ============================================================
    
    /**
     * Get the effective subject (mapped subject if available, otherwise original)
     */
    public Subject getEffectiveSubject() {
        return mappedSubject != null ? mappedSubject : subject;
    }
    
    /**
     * Check if mapping confidence is high (>= 80%)
     */
    public boolean hasHighMappingConfidence() {
        return mappingConfidence != null && 
               mappingConfidence.compareTo(new BigDecimal("80.00")) >= 0;
    }
    
    /**
     * Check if this topic has been manually reviewed/approved
     */
    public boolean isMappingApproved() {
        return mappedSubject != null;
    }
}

File: python-service/app/models/individual_lesson_topic.py
"""
app/models/individual_lesson_topic.py
SQLAlchemy model for individual lesson topics extracted from schemes
"""
from sqlalchemy import Column, Integer, String, DateTime, ForeignKey, Text, Float
from sqlalchemy.sql import func
from app.models.base import Base


class IndividualLessonTopic(Base):
    """Stores extracted lesson topics from scheme of work for INDIVIDUAL students"""
    
    __tablename__ = 'individual_lesson_topics'
    __table_args__ = {'schema': 'academic'}
    
    # Primary key
    id = Column(Integer, primary_key=True, index=True)
    
    # Foreign keys
    student_profile_id = Column(Integer, nullable=False, index=True)
    scheme_id = Column(Integer, ForeignKey('academic.individual_student_schemes.id'), nullable=False)
    subject_id = Column(Integer, nullable=False, index=True)
    term_id = Column(Integer, nullable=True)
    
    # Topic details
    topic_title = Column(String(255), nullable=False)
    description = Column(Text, nullable=True)
    week_number = Column(Integer, nullable=True)
    
    # Subject mapping (AI-generated)
    mapped_subject_id = Column(Integer, nullable=True)  # Platform subject after mapping
    mapping_confidence = Column(Float, nullable=True)  # 0.00 to 100.00
    
    # Learning resources
    file_name = Column(String(255), nullable=True)
    file_url = Column(String(500), nullable=True)
    
    # Timestamps
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())
    
    def __repr__(self):
        return f"<IndividualLessonTopic(id={self.id}, student={self.student_profile_id}, week={self.week_number}, topic='{self.topic_title[:30]}')>"
    
    # Helper methods matching Java entity
    def has_high_mapping_confidence(self):
        """Check if mapping confidence is high (>= 80%)"""
        return self.mapping_confidence is not None and self.mapping_confidence >= 80.0
    
    def is_mapping_approved(self):
        """Check if this topic has been manually reviewed/approved"""
        return self.mapped_subject_id is not None